### ISSUE TITLE
[2021 ICML Spotlight] Barlow Twins: Self-Supervised Learning via Redundancy Reduction

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#54]
기존의 SSL은 large batch size를 요구하며 asymmetric learning update 등 까다로운 학습 방법을 가지고 있었음

해당 논문은 새로운 loss function을 제시하며 상대적으로 적은 batch size와 symmetric한 학습 방법으로도 충분히 괜찮은 성능을 보인 논문임

Method, results, ablations만 설명

### 1.1. Description of Barlow Twins
<img width="713" alt="image" src="https://user-images.githubusercontent.com/49643709/156317075-3a973104-1ab2-458a-b66e-c63aa8b8d476.png">
<img width="674" alt="image" src="https://user-images.githubusercontent.com/49643709/156318860-7b490c33-5107-4420-9125-ee626713bb11.png">
<img width="675" alt="image" src="https://user-images.githubusercontent.com/49643709/156318615-3cfffffe-d866-4bdc-b6ae-1924430f6846.png">

해당 논문은 새로운 loss function을 제시함

Invariance term을 통해 embedding이 적용된 distortion에 invariant하게 만들어주며 redundancy reduction term을 통해 embedding의 vector component를 decorrelate하게 만듬

람다 값을 조정하여 invariance term과 redundancy reduction term의 중요성을 trade-off
(positive constant)

Batch dimension으로 normalize를 하는 것과 loss function을 제외하고 기존의 contrastive learning based SSL과 비슷함

#### 1.2.1. Image augmentations

- Random crop and resize to 224 * 224 (always applied)
- Random horizontal flipping
- Random color jittering
- Random convert to gray scale
- Random Gaussian blurring (prob is different between 2 distorted views)
- Random solarization (prob is different between 2 distorted views)

Augmentation parameter는 BYOL과 같음

#### 1.2.2. Architecture

- Encoder : ResNet 50 wtihout final fc layer
- Projector : 3 fc layers (2048 - 8192 fc + BatchNorm + ReLU + 8192 - 8192 fc + BatchNorm + ReLU + 8192 - 8192 fc)

- Representation : encoder output
- Embedding : projector output

#### 1.2.3. Optimization

- Lambda on loss function : 0.005
- LARS optimizer
- 1000 epoch
- 2048 batch size (works well with batches as small as 256)
- Learning rate 0.2 for weight, 0.0048 for bias and batch norm parameters
- Learning rate warm-up for 10 epochs
- Cosine decay schedule with factor of 1000
- Weight decay 0.0000015
- Bias, batch norm parameters are excluded from LARS adaptation, weight decay

32 V100 GPU로 124시간 학습

### 2.1. Linear and Semi-Supervised Evaluations on ImageNet
<img src='https://user-images.githubusercontent.com/49643709/156321115-a93c4ae0-8041-42cd-8ffe-d9f6eadd42a3.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/156321161-95ae11c7-da5b-4d5f-88d0-e333e152a284.png' width=80% height=80%>

### 2.2. Transfer to other datasets and tasks
<img src='https://user-images.githubusercontent.com/49643709/156321375-3ac54dde-a30c-4680-a80c-4ea3f2bc62e1.png' width=80% height=80%>
<img src='https://user-images.githubusercontent.com/49643709/156321413-bf1640ec-35f7-47cb-a4d8-72e42a3b8788.png' width=80% height=80%>

### 3.1. Loss Function Ablations
<img src='https://user-images.githubusercontent.com/49643709/156322253-326bf08a-edb1-461b-a9b6-2690cafcda8b.png' width=80% height=80%>

### 3.2. Robustness to Batch Size
<img width="666" alt="image" src="https://user-images.githubusercontent.com/49643709/156477466-ad36a354-7de4-4085-ae22-4af8ab7fcc3c.png">

Small batch size에서도 robust한 성능을 냄

### 3.3. Effect of Removing Augmentations
<img width="687" alt="image" src="https://user-images.githubusercontent.com/49643709/156477544-70512878-cc3d-4445-aab2-5010a1fde2c0.png">

Data augmentation에 대해서 robust하지 않음

### 3.4. Projector Network Depth & Width
<img width="698" alt="image" src="https://user-images.githubusercontent.com/49643709/156477864-b8797ece-6a75-40bf-89bc-ca619a4525db.png">

Projector의 layer가 deep할수록, projector output의 dimension이 클수록 성능 향상
(3 layer 이상 stack하면 saturation)

### 3.5. Breaking Symmetry
<img width="666" alt="image" src="https://user-images.githubusercontent.com/49643709/156478283-322660ce-fad4-414f-b5ea-2d4a62e5e5f6.png">

기존의 trivial solution을 방지하기 위한 symmetry-breaking mechanism을 적용하면 오히려 성능이 감소

### 3.6. BYOL with a larger projector/predictor/embedding
<img width="701" alt="image" src="https://user-images.githubusercontent.com/49643709/156478444-f1f35a91-7b7d-4956-82d1-ef7050b64e5b.png">

BYOL과 fair comparison을 하기 위해 실험 진행

### 3.7. Sensitivity to Lambda
<img width="683" alt="image" src="https://user-images.githubusercontent.com/49643709/156478626-2c8bad34-d7d0-4e12-8c73-37e1608ac31f.png">

Not very sensitive to lambda

### 3.8. InfoNCE and Barlow Twins

결국 기존의 contrastive learning과는 다른 특징이 존재
(small batch size, no asymmetric mechanism, large projector dimension, ...)

해당 이유로 objective function의 차이라고 분석함

InfoNCE는 embedding distribution의 entropy에 대한 non-parametric estimation로 해석할 수 있기에, 차원의 저주에 취약함

저차원에서만 신뢰성 있게 estimate할 수 있기에, embedding dimension이 작아야함

또한 InfoNCE에서의 hyperparameter인 temperature는 entropy의 non-parametric kernel density estimation에서의 kernel width로 해석할 수 있고, 이는 실질적으로 batch 내의 hardest negative에 대해서 relative importance를 weight하기에 large batch가 필요함

그에 반해 Barlow Twins는 Gaussian parameterization 내의 embedding distribution에 대한 proxy entropy estimator라고 해석할 수 있음

Simplified parameterization으로 인해 small batch를 사용해도 되고 embedding dimension이 커도 됨
